metricbeat/module/windows/perfmon/pdh_query_windows.go

```diff
@@ -0,0 +1,200 @@
+		}
+	} else {
+		instanceName = counter.InstanceName
+	}
+	h, err := PdhAddCounter(q.handle, counterPath, 0)
+	if err != nil {
+		return err
+	}
+
+	q.counters[counterPath] = &Counter{
+		handle:       h,
+		instanceName: instanceName,
+		format:       getPDHFormat(counter.Format),
+	}
+	return nil
+}
+
+// CollectData collects the value for all counters in the query.
+func (q *Query) CollectData() error {
+	return PdhCollectQueryData(q.handle)
+}
+
+// GetFormattedCounterValues returns an array of formatted values for a query.
+func (q *Query) GetFormattedCounterValues() (map[string][]CounterValue, error) {
+	if q.counters == nil || len(q.counters) == 0 {
+		return nil, errors.New("no counter list found")
+	}
+	rtn := make(map[string][]CounterValue, len(q.counters))
+	for path, counter := range q.counters {
+		rtn[path] = append(rtn[path], getCounterValue(counter))
+	}
+	return rtn, nil
+}
+
+// ExpandWildCardPath  examines local computer and returns those counter paths that match the given counter path which contains wildcard characters.
+func (q *Query) ExpandWildCardPath(wildCardPath string) ([]string, error) {
+	if wildCardPath == "" {
+		return nil, errors.New("no query path given")
+	}
+
+	expdPaths, err := PdhExpandWildCardPath(wildCardPath)
+	if err != nil {
+		return nil, err
+	}
+	return UTF16ToStringArray(expdPaths), nil
+}
+
+// Close closes the query and all of its counters.
+func (q *Query) Close() error {
+	return PdhCloseQuery(q.handle)
+}
+
+// matchInstanceName will check first for instance and then for any objects names.
+func matchInstanceName(counterPath string) (string, error) {
+	matches := instanceNameRegexp.FindStringSubmatch(counterPath)
+	if len(matches) != 2 {
+		matches = objectNameRegexp.FindStringSubmatch(counterPath)
+	}
+	if len(matches) == 2 {
+		return matches[1], nil
+	}
+	return "", errors.New("query doesn't contain an instance name. In this case you have to define 'instance_name'")
+}
+
+// getCounterValue will retrieve the counter value based on the format applied in the config options
+func getCounterValue(counter *Counter) CounterValue {
+	counterValue := CounterValue{Instance: counter.instanceName}
+	switch counter.format {
+	case PdhFmtLong:
+		_, value, err := PdhGetFormattedCounterValueLong(counter.handle)
+		if err != nil {
+			return CounterValue{Err: err}
+		}
+		counterValue.Measurement = value.Value
+	case PdhFmtLarge:
+		_, value, err := PdhGetFormattedCounterValueLarge(counter.handle)
+		if err != nil {
+			return CounterValue{Err: err}
+		}
+		counterValue.Measurement = value.Value
+	default:
+		_, value, err := PdhGetFormattedCounterValueDouble(counter.handle)
+		if err != nil {
+			return CounterValue{Err: err}
+		}
+		counterValue.Measurement = value.Value
+	}
+	return counterValue
+}
+
+// getPDHFormat calculates data pdhformat.
+func getPDHFormat(format string) PdhCounterFormat {
+	switch format {
+	case "long":
+		return PdhFmtLong
+	case "large":
+		return PdhFmtLarge
+	default:
```

Comment:
Ok, I see that we check for proper values when reading the config, but I'd still feel better if we checked for valid values here and in `getCounterValue` :grimacing:

Reply:
I have added an extra check when retrieving the counter values, thanks for the feedback.